src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -204,9 +205,24 @@ private void AddCompilationRoots()
                 AddCompilationRootsForRuntimeExports(module);
            }
 
+            AddCompilationRootsForBootstrap();
+
             AddCompilationRootsForRuntimeExports((EcmaModule)_typeSystemContext.SystemModule);
         }
 
+        private void AddCompilationRootsForBootstrap()
+        {
+            if (Options.IsCppCodeGen)
+                return;
+
+            if (_mainMethod == null)
+                return;
+
+            var owningType = _typeSystemContext.GetWellKnownType(WellKnownType.Object);
```

Comment:
(If you do that, you can also make `IsModuleType` on `MetadataType` non-virtual and implement it right there...)